### ISSUE TITLE
Likely namespacing bug with unions, collect, and list comprehension being used together Neo4j 4.1.1

### ISSUE THREAD

**Issue by MichaelCDormann** [neo4j/neo4j#12599]
I discovered that using unions, collect, and list comprehension together a certain way causes this error: `There was a cycle in names: Set(    val@0@0,   AGGREGATION297). This is likely a namespacing bug.`

**Neo4j Version:** 4.1.1
**Operating System:** Red Hat Enterprise Linux:7.8:GA:server

### Steps to reproduce
Run this query in the Neo4j browser for a 4.1.1 cluster.
```
CALL {
  WITH 'testval' AS val RETURN val
  UNION 
  WITH 'anotherval' AS val RETURN val
}
RETURN val, [] AS thisBreaks
UNION
CALL {
    WITH 'valval' AS val RETURN val
    UNION
    WITH 'valvalval' AS val RETURN val
}
WITH val, [v IN collect(val) WHERE v = 'valval' | v] AS thisBreaks
RETURN val, thisBreaks
```

### Expected behavior
A result is returned.

### Actual behavior
An error like `There was a cycle in names: Set(    val@0@0,   AGGREGATION225). This is likely a namespacing bug.` occurs.

### More examples
The error doesn't happen if you break up the collect and the list comprehension into two steps. The query below returns a result.
```
CALL {
  WITH 'testval' AS val RETURN val
  UNION 
  WITH 'anotherval' AS val RETURN val
}
RETURN val, [] AS thisBreaks
UNION
CALL {
    WITH 'valval' AS val RETURN val
    UNION
    WITH 'valvalval' AS val RETURN val
}
WITH val, collect(val) as col
WITH val, [v IN col WHERE v = 'valval' | v] AS thisBreaks
RETURN val, thisBreaks
```

The error also doesn't happen if you remove the outer union. The query below returns a result.
```
CALL {
    WITH 'valval' AS val RETURN val
    UNION
    WITH 'valvalval' AS val RETURN val
}
WITH val, [v IN collect(val) WHERE v = 'valval' | v] AS thisBreaks
RETURN val, thisBreaks
```

The error also doesn't happen without the subqueries. This query below returns a result.
```
WITH 'testval' AS val
RETURN val, [] AS thisBreaks
UNION
WITH 'valvalval' AS val
WITH val, [v IN collect(val) WHERE v = 'valval' | v] AS thisBreaks
RETURN val, thisBreaks
```

**Comment by mnd999** [comment]
Thanks for the report, we'll take a look into it.

**Comment by InverseFalcon** [comment]
Note that you can avoid this by doing the collect() aggregation outside of the pattern comprehension.

```
explain
CALL {
  WITH 'testval' AS val RETURN val
  UNION 
  WITH 'anotherval' AS val RETURN val
}
RETURN val, [] AS thisBreaks
UNION
CALL {
    WITH 'valval' AS val RETURN val
    UNION
    WITH 'valvalval' AS val RETURN val
}
WITH val, collect(val) as vall
WITH val, [v IN vall WHERE v = 'valval' | v] AS thisBreaks
RETURN val, thisBreaks
```

In general, it's not a good idea to perform an aggregation within a pattern comprehension like this, as it can make it confusing or unclear what the grouping key is for the aggregation. 

The bug should and will be fixed, but still the conditions to trigger it violate best practice.

**Comment by mnd999** [comment]
Fixed by https://github.com/neo4j/neo4j/commit/ee8e0601c5e0e0004700c1e03310e6fc17ac5370 
This fix will be included in the next 4.1.x bugfix release
